Fix experimental_indexers `importPath` being ignored

svallory opened the pull request:
Closes #25554

## What I did

Added a simple "or" to try to use indexers.importPath but fallback to the original file path.

### Couple of things to note

1. `importPath` is currently required. I think it is safe to make it optional with the default being the path of the file being indexed. I haven't added a test case because of this.
2. Contribution guide didn't mention anything about bumping the version, so I didn't

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

svallory commented:
@JReinhold when you have some time, please take a look at the questions I had

JReinhold commented on code/lib/core-server/src/utils/StoryIndexGenerator.ts:
```diff
@@ -322,7 +322,8 @@ export class StoryIndexGenerator {
           name,
           title,
-          importPath,
+          // We must use `||` instead of `??` since "" (empty string) is not a valid import path
```
When would an indexer return an empty string here?

I'm okay with adding support for undefined and just fallback to the original as you did here, but then we need to make it optional in the types too.

svallory replied:
Let's say there's a function that is generating the import path, or the user thinks that an empty string will prevent the file from being imported... 

On a second thought though I think it's better to throw an "Invalid importPath defined by the indexer".

Am I free to change the types and make the importPath optional?

svallory replied:
To speed up things, I went ahead and changed the type. It seemed like the right choice

svallory commented:
Hey @JReinhold, I got my storybook sandbox working yesterday to test this PR and there was a piece missing. The `builder-vite` package also needed to be updated to handle the `importPath` correctly.

Initially I thought of importing `StoryIndexGenerator` from `core-server`  but I quickly ran into issues related to esm vs cjs. So instead I assumed that a custom `importPath` provided by an indexer will **ALWAYS** be virtual. I think that's 

Similar problems may arise for other builders. Since I'm still learning about the inner workings of Storybook, it would be great if someone from the core team could take at this. It would take me quite some time to verify since I would still need to setup sandboxes for each builder.

If someone finds and issue with a builder and points me to the right location, I can quickly fix other builders applying the same logic.

After I've received an approval for the implementation, I'll add tests, an `importPath` validation, and a note to the docs to inform of the `importPath` requirements. For example, that the `importPath` needs to be either the original path or a virtual module path following vite/rollup convention (if that's what we decide)

svallory commented:
crap... @ndelangen I wanted to pull and did the opposite 🤦🏻 and I just realized it's too easy to do a force push with Lazygit.

I noticed there was a commit from you before I did this atrocity. Was it just a rebase on top of `latest-release`?

svallory commented:
I've rebased the branch on top of `latest-release` and added a commit that fixes the `ExtractPresetType` (renamed to `UnwrapPresetType`) and (hopefully) all the build issues

svallory commented:
Hey @JReinhold or @valentinpalkovic, it looks to me that what we need now is to update the snapshots due to the change of the import path. But I think it is better if someone from the core team take a look at it

JReinhold commented:
> So instead I assumed that a custom importPath provided by an indexer will ALWAYS be virtual.

Can you tell me more about this, why do you assume this?

JReinhold commented on code/builders/builder-manager/src/index.ts:
```diff
@@ -33,7 +33,7 @@ export const getConfig: ManagerBuilder['getConfig'] = async (options) => {
     safeResolve(join(options.configDir, 'manager')),
     getTemplatePath('addon.tsconfig.json'),
-    options.presets.apply<Record<string, string>>('env'),
+    options.presets.apply('env'),
```
Why was this necessary?

svallory replied:
Type of presets is now automatically detected

JReinhold commented on code/builders/builder-manager/src/utils/data.ts:
```diff
@@ -8,14 +8,14 @@ import { executor, getConfig } from '../index';
-  const features = options.presets.apply<Record<string, string | boolean>>('features');
-  const logLevel = options.presets.apply<string>('logLevel');
-  const title = options.presets.apply<string>('title');
-  const docsOptions = options.presets.apply<DocsOptions>('docs', {});
+  const features = options.presets.apply('features');
+  const logLevel = options.presets.apply('logLevel');
+  const title = options.presets.apply('title');
+  const docsOptions = options.presets.apply('docs', {});
   const template = readTemplate('template.ejs');
-  const customHead = options.presets.apply<string>('managerHead');
+  const customHead = options.presets.apply('managerHead');
```
seems like we're loosening the types here, why?

svallory replied:
It's actually the contrary. The type parameter would accept and allow the caller to override the actual type of the property.

JReinhold reviewed (commented):
I'm seeing lots of changes here unrelated to `importPath` - mostly preset handling and typing. Are they a result of a bad rebase/merge or are all these changes intentional?

Either way I think we should keep them out of this PR.

svallory commented:
> Can you tell me more about this, why do you assume this?

Instead of assume, I should have said "I propose"  that we restrict paths generated by custom indexers to virtual modules.  I had an issue with builder-vite. I don't remember all the details anymore, but looking at the code I think you'll be able to figure out the implications. Restricting it to virtual modules we can do this:  https://github.com/storybookjs/storybook/pull/26010/files#diff-a4ade24974a8c7f082261c73f22be07ec54ad61a3f9683b20035955a50fa1a8bR55-R57

My rationale is that custom indexers are pre-processing a file before handing it over to Storybook, otherwise a normal import would suffice. That's the definition of a virtual module and there's a naming standard in vite that comes from rollup for that. In a way, using virtual paths is a way to ensure the stories file will be plucked from the normal processing pipeline into a custom one.

Let me know if this is acceptable or not and what would be a better solution.

svallory commented:
> I'm seeing lots of changes here unrelated to `importPath` - mostly preset handling and typing. Are they a result of a bad rebase/merge or are all these changes intentional?
> 
> Either way I think we should keep them out of this PR.

The changes were actually semi-intentional, and after my last commits I also started to think it should go into a separate PR. Here's the history behind it:

As a newcomer to Storybook, I didn't know which type I needed to set in the type parameter, so I went to investigate and this was [the definition of the `apply` function](https://github.com/storybookjs/storybook/pull/26010/files#diff-8ddce16fd5ff342d3a2ab4bd13455d2504679a0c222f7d68fafd1d9132a29c1eL79-L82):

```ts
export interface Presets {
 ...
  apply(extension: 'refs', config?: [], args?: any): Promise<unknown>;
  apply(extension: 'core', config?: {}, args?: any): Promise<CoreConfig>;
  apply(extension: 'build', config?: {}, args?: any): Promise<StorybookConfig['build']>;
  apply<T>(extension: string, config?: T, args?: unknown): Promise<T>;
}
```

That has 2 issues:

1. The config parameter of explicitly defined `extension`s was never typed
2. I could call apply with anything I wanted if using `apply<T>`

So I though "let me improve this type to have a strong type check and give me autocomplete".  So [I changed it to this](https://github.com/storybookjs/storybook/pull/26010/files#diff-8ddce16fd5ff342d3a2ab4bd13455d2504679a0c222f7d68fafd1d9132a29c1eR90-R100):

```
apply<TKey extends keyof AllSupportedPresets>(
    extension: TKey,
    config?: Partial<UnwrapPresetType<TKey>> | null,
    args?: any
  ): Promise<UnwrapPresetType<TKey>>;

  apply<TReturn = never>(
    extension: string,
    config?: Partial<TReturn> | null,
    args?: any
  ): Promise<TReturn>;
```

That type has stronger type checking for properties defined in `StorybookConfig` and a slightly better `apply<T>`. I imagine that since everything was supposedly working, it wouldn't have a big impact and could be adopted gradually. 

An initial run of the tests seemed to confirm that, but I later discovered via multiple CI failures that the type checking gives a lot of false passes.

The places where the type parameter was removed are places that are now relying on the types automatically detected by the apply function. In a few places that add properties not present in the `StorybookConfig` a type parameter had to be added. In some places, the typing was wrong and had to be fixed.

I only did this because Storybook codebase is huge, and I needed a way to explore the options and ensure I wasn't doing a mistake. In the previous `apply` config would always be `{}` or `[], so passing a `build` config to the `core` preset would not be prevented by the parameter type.

Again, maybe this should go into a different PR, it's your call, but I definitely recommend that we keep this change.

svallory commented:
Guys, has this been fixed in Storybook 8?

If not, I'll rebase and cleanup the PR.

I need this to finish the marko-stories package :/

JReinhold commented:
@svallory This hasn't been fixed in 8.0 yet. If you rebase this PR on `next` I should be able to help get it over the finish line this week. 🙏

> My rationale is that custom indexers are pre-processing a file before handing it over to Storybook, otherwise a normal import would suffice.

I agree with this and I can't immediately see a use case where users would use a custom `importPath` and _not_ use virtual modules, except if they wanted to generate the CSF files to the file system for some reason. I'm fine with this limitation of the API for now, we can expand it later if someone comes up with good use cases for it.

I must admit I don't fully understand the `importFn` code changes right now or why they are necessary, but maybe diving more into it later will make it click for me.

Thank you for the thorough explanation of the `apply` type changes, that makes a lot of sense. However given they are so wide-spread I think they should still be in a separate PR.

svallory commented:
> I must admit I don't fully understand the `importFn` code changes right now or why they are necessary, but maybe diving more into it later will make it click for me.

I don't know exactly why it is necessary but I have two guesses:

- It converts normal imports to async function imports so we don't need to import all story files at build time
- It deals with the vite limitation of not supporting dynamic import paths
 
In any case, I just went with the flow there and made sure the function also supported virtual story paths



> Thank you for the thorough explanation of the `apply` type changes, that makes a lot of sense. However given they are so wide-spread I think they should still be in a separate PR.

Yep, agreed! I wasn't expecting such a big impact when I implemented it.

I'm pretty busy this week, but I'll find some time on the weekend to work on it

valentinpalkovic commented:
Hi @svallory 

Something went wrong on this branch. I guess the base branch wasn't `next`, but `latest-release`, which now leads to a bunch of conflicts and a huge diff. I tried cherry-picking your commits into a new branch, but I wasn't sure about the conflicts. Could you create another PR with `next` as the base branch?

svallory commented:
> Hi @svallory 
> 
> 
> 
> Something went wrong on this branch. I guess the base branch wasn't `next`, but `latest-release`, which now leads to a bunch of conflicts and a huge diff. I tried cherry-picking your commits into a new branch, but I wasn't sure about the conflicts. Could you create another PR with `next` as the base branch?

I've been wanting to do it for a while but work has been eating even my weekends.

I'll try to find some time in the next days, but if not, I'll finish it next weekend.

svallory commented:
@valentinpalkovic I rebased my branch on top of next and just pushed the fix for indexer's importPath. I also updated the affect tests snapshots.

Note:

- the tests that are failing were already failing on a clean checkout of the `next` branch
- the `importPath` returned by indexers is an absolute path (I don't know if that is an issue). I think it must be used as is since importers need to be able to return virtual paths and it would be very hard to figure out if the path can be made relative to the project/source root or not

svallory commented:
Btw, @valentinpalkovic I accidentally committed a new sandbox template for Marko (writing stories in Marko is the whole reason I'm sending this PR).

I'll remove it in the next commit.

Would a PR to add Marko JS as a sandbox template and fully supported framework acceptable?

svallory committed:
fix: exprerimental indexers returned path being ignored

JReinhold reviewed (changes requested):
Great cleanup, much easier to review now. 😅 Still a couple of questions and feedback points to touch on.

Could you fill out the "Manual Testing" section in the PR, with steps on how to test this? Just something minimal, with a custom indexer returning a custom virtual `importPath`, and a very minimal Vite plugin that adds that virtual module with some CSF content in it.

I'm sorry if it feels like I'm nitpicking here, but you're touching on some of the most essential code in Storybook, and breaking it could cause issues for many users. We'll get there! 😊

svallory commented:
> I'm sorry if it feels like I'm nitpicking here, but you're touching on some of the most essential code in Storybook, and breaking it could cause issues for many users. We'll get there! 😊

Hey, @JReinhold, don't worry about it! I wanna get this right! ;)

I was actually trying to get a sandbox for Marko to test this fix, but I ran into to a lot of issues while trying to do that. 

Can you help me get that up and running? I think it would become an awesome example of customization since it will touch on all that one can do using custom indexers. In trying to fix it, I started adding Marko alongside all the other frameworks in every place I could find, and it feels like I'm close now.

I've pushed my WIP to the `feature/marko-framework` of my fork and [created a PR to this branch](https://github.com/svallory/storybook/pull/1) to make it easier for you to check what changes I did. The `marko-stories` code isn't there, I added the [@storybookjs/marko](/storybookjs/marko) repo as a submodule in `code/marko` to make it easier to create a sandbox.

It would be great to use the add-on I'm already creating as the manual test. But if this is all too much work, let me know and I'll come up with something else.

I intend to work on this tomorrow to address all the comments you've made.

JReinhold commented:
> I've pushed my WIP to the `feature/marko-framework` of my fork and [created a PR to this branch](https://github.com/svallory/storybook/pull/1) to make it easier for you to check what changes I did. The `marko-stories` code isn't there, I added the [@storybookjs/marko](/storybookjs/marko) repo as a submodule in `code/marko` to make it easier to create a sandbox.

I quickly looked through it, looks straightforward and you are touching the right files, but I don't think the git submodule idea will hold. I suggest you draw inspiration from the Solid and Qwik integrations, which are both examples of frameworks+renderers that are initable and sandboxable but are actually not in the monorepo. I think there are some limitations with this, but for now this the best we can do.

https://github.com/storybookjs/storybook/blob/next/code/lib/cli/src/sandbox-templates.ts/#L492-L504

https://github.com/storybookjs/storybook/blob/next/code/lib/cli/src/generators/QWIK/index.ts

https://github.com/storybookjs/storybook/blob/next/code/lib/cli/src/project_types.ts/#L23-L26

> It would be great to use the add-on I'm already creating as the manual test. But if this is all too much work, let me know and I'll come up with something else.

We use the Manual Test section to manually verify things work before we to minor/major releases (in about 2 months). The best way to do that are with minimal steps, similar to a minimal reproduction. It's just more fool proof if the manual steps are _minimal_, and don't depend on bigger things. If things fail it could be impossible to tell if they fail because of some bug in the Marko integration or because there's an issue with the import paths, and I want to avoid that confusion if I can.

svallory commented:
> I quickly looked through it, looks straightforward and you are touching the right files

That's great! I was concerned I was doing way too much just to support sandboxing.

> I don't think the git submodule idea will hold

That's just to make my exploration easier. I didn't intend to propose it. 

>  suggest you draw inspiration from the Solid and Qwik integrations, which are both examples of frameworks+renderers that are initable and sandboxable

Great! That makes it way easier.

> The best way to do that are with minimal steps

I agree. This should not be too hard. In fact, I think I should also add a unit test for indexers returning virtual paths. And a note to the docs about it (since there's a recommendation from vite/rollup regarding their syntax). I think I'll take inspiration on the JSON stories from the docs and make an example that embeds the stories file content in a virtual path using base64. That will eliminate the fs loading and parsing in vite which can be cumbersome to do at runtime.

One more thing...

I noticed your comment about the dynamic snapshot. That was the easiest path to make sure everything was working. The only way I see out of that is to make the path relative again. We can use the  [`StoryIndexGenerator.options.workingDir`](https://github.com/svallory/storybook/blob/feature/marko-framework/code/lib/core-server/src/utils/StoryIndexGenerator.ts#L47) for that.

Also, we can do this in the generator's [extractStories() function](https://github.com/svallory/storybook/blob/feature/marko-framework/code/lib/core-server/src/utils/StoryIndexGenerator.ts#L318), but that wouldn't work for all tests or all paths.

The only way to ensure the removal of machine-specific paths in the return would be to process the generator's result in the tests before comparing it to the snapshot.

Please let me know how to proceed. I'll work on the final touches tonight/tomorrow

svallory commented on code/builders/builder-vite/src/codegen-importfn-script.ts:
```diff
@@ -24,12 +25,24 @@ function toImportPath(relativePath: string) {
+
+    return [
+      `  '${toImportPath(relativePath)}': async () => import('/@fs/${file}'),`,
+      `  '${file}': async () => import('/@fs/${file}')`,
```
That line is adding an alias for importing `file` using its absolute path. It is necessary because now storybook may try to import a file using either a relative or absolute path. Before it always used a relative path.

The `/@fs/` prefix will make Vite skip any path resolution and go straight to the file system to read the file.

It also avoids conflicts with any virtual path handler configured in Vite, which can be trigger by valid file system paths.

We should not do anything regarding virtual paths here as each vite plugin may choose to handle its virtual paths at a different moment/hook in vite's lifecycle.

tmeasday replied:
I don't know if we should be allowing importing absolute paths? Is this change also required for webpack?

svallory commented on code/lib/telemetry/src/storybook-metadata.test.ts:
```diff
@@ -348,8 +348,8 @@ describe('storybook-metadata', () => {
           refs: {
-            a: { title: '', url: '' },
-            b: { title: '', url: '' },
+            a: { id: '', version: '', title: '', url: '' },
+            b: { id: '', version: '', title: '', url: '' },
```
Nothing. I have no idea how/why these changes got there.

tmeasday commented on code/lib/core-server/src/utils/StoryIndexGenerator.test.ts:
```diff
@@ -76,7 +78,7 @@ describe('StoryIndexGenerator', () => {
               "a--story-one": {
                 "id": "a--story-one",
-                "importPath": "./src/A.stories.js",
+                "importPath": "${path.join(workingDir, './src/A.stories.js')}",
```
I don't understand why there would be absolute paths in the index. This seems like it shouldn't happen, and is a security issue. Perhaps we need to relativize the path returned from the indexer?

svallory commented:
@tmeasday 

> I don't understand why there would be absolute paths in the index.

The reason it is there is that Storybook passes an absolute path to the importer. Since the test importers simply return the path, the absolute path gets into the index.

> This seems like it shouldn't happen, and is a security issue. 

That was my first reaction too, but if you think about it, there's nothing preventing the indexer code of reading the entire filesystem if it wants to. AFAIK, unless Storybook implements a require/import handler to override node's fs module (if that's even possible), that will remain to be the case, right? Or is there something I'm missing here?

I do feel it is kinda odd to deal with absolute paths, in a web app. But may get complicated on that front when you factor in all the variables involved. If we do relativize the path, what will be used as the base for import? The project root, the build/web root, the custom indexer file? When the bundler plugin for the indexer gets the file, will it get the relative or the absolute path? If it's a relative path, can we make sure the `cwd` when the indexer's bundler plugin is the same as the base path used to relativize the story path?

What I'm trying to point out here is that there are a lot of things to think about when forcibly modifying the indexer's return. This is a decision that the Storybook team will need to discuss and take very carefully.

I think it is outside of the scope of this PR though. The goal here is just to make indexers work as advertised in the docs. Right now, what is documented in the site does not work. It never has. I hope the fact the the indexers are `experimental` is enough to allow us to move on and change the path behavior (if that's the decision) in another PR

I'll be pushing the fixes for all other issues later today!

tmeasday commented:
> If we do relativize the path, what will be used as the base for import? The project root, the build/web root, the custom indexer file?

The CWD is what we have been using up until now.

Basically the import path is going to be relative to the virtual module that defines the `importFn` and does the actual import. And we've been adding that virtual module at the CWD ([here's how webpack does it](https://github.com/storybookjs/storybook/blob/dcd5c9689a5d782f5d22ca3b8b1ff97b8c9855ea/code/builders/builder-webpack5/src/preview/virtual-module-mapping.ts#L49) - I'm less sure about Vite).

> I think it is outside of the scope of this PR though. The goal here is just to make indexers work as advertised in the docs. Right now, what is documented in the site does not work. It never has. I hope the fact the the indexers are experimental is enough to allow us to move on and change the path behavior (if that's the decision) in another PR

I'm not really too concerned about the behaviour of customer (experimental as you say) indexers, I am concerned about the default behaviour in Storybook as exhibited by the tests. Am I off base about that?

svallory commented:
> it isn't the thing doing the importing so should have to fit with the existing `import()` behavior.

I think you are right. As long as we only have imports of real files in the `importFn` making the paths relative should not cause any issues. At least not now, since there's no plugin relying on that path. So it is better to make this change now, before people start using it.

> I am concerned about the default behaviour in Storybook as exhibited by the tests. Am I off base about that?

Not at all! What I meant was that Storybook isn't introducing any security risks by exposing the absolute path to the indexer. But that's because I assumed that the code generated for those imports was only used in `dev` mode. If that's not the case, or may not be the case in the future, then we would be exposing paths to the web that may indeed aid an attacker.

tmeasday commented:
> If that's not the case, or may not be the case in the future, then we would be exposing paths to the web that may indeed aid an attacker.

Right, so yes, the `index.json` ("stories json") is produced in production builds of SB so it would be a big problem to have absolute paths in there. Honestly I think folks would find it pretty disconcerting if they were in there in dev mode too (and as it happens the `index.json` should be the same in both modes anyway).

svallory committed:
[wip] fix: removing absolute paths

tmeasday commented:
> Check this example from `@storybook/builder-vite`
>   `export const virtualStoriesFile = '/virtual:/@storybook/builder-vite/storybook-stories.js'`

@svallory but that path wouldn't end up in the `index.json` right? Is there a scenario you can think of where an indexer would ever create a virtual path for a story file? If not, I'd say we could just not allow it and throw if you pass a path that isn't either a simple relative or absolute path.

JReinhold commented:
> Is there a scenario you can think of where an indexer would ever create a virtual path for a story file? 

I believe that's actually the point of this whole exercise @tmeasday , and the only use case for indexers to specify their own `importPath`. (We explored this topic further up in the PR but I understand it's hard to keep up with this)

If you have an input file - `allFixtures.json` that you want to split up into multiple components, the best way to do that is to generate a virtual module for each component, that the indexer then sets as the `importPath`.
I'm not sure there's any other use case for setting an `importPath` different from the original

> We can enforce a specific prefix for virtual paths (traditionally `virtual:pkg-name`), and informed the user of the requirement on every import error "in case they are trying to use a virtual path.

Vite already establishes a convention for this with the `virtual:` and `\0virtual:` prefix and I think we should just lean into that. If the `importPath` starts with either fo those, consider it a virtual module. If not, assume it's a file path and convert it to a relative path if needed.

https://vitejs.dev/guide/api-plugin#virtual-modules-convention

tmeasday commented:
Ok, thanks for filling me in (and sorry for not reading the whole thread). That approach sound sensible to me!

dunklesToast commented:
just came across this thread because i think this would also solve an issue i am currently facing. would it be possible for somebody from the @storybookjs/core team to create a canary release for pr so i can check if this works for my usecase as well or if i need to dig somewhere else?

nahtnam commented:
Same here, happy to test a canary. Thank you @svallory for working on this

dunklesToast commented:
I'd love to take over but after reading the thread I am not 100% sure whats left to be done here? Probably fixing the TypeScript types that break the build right now and some manual testing?

svallory commented:
> I'd love to take over but after reading the thread I am not 100% sure whats left to be done here? Probably fixing the TypeScript types that break the build right now and some manual testing?

Hey @dunklesToast, that's great! There isn't much left to do. IIRC, all that is left is:

- [ ] Make sure absolute paths are turned into relative ones
  This will affect [StoryIndexGenerator.ts](https://github.com/storybookjs/storybook/pull/26010/files#diff-07c9399482e8dcbe6a2f17adee5780b4af5f7cead706e5b07d28346487378319) and maybe can be entirely solved there. 
  For context, check these comments:
  - [#issuecomment-2135635954](https://github.com/storybookjs/storybook/pull/26010#issuecomment-2135635954)
  - [#issuecomment-2136405529](https://github.com/storybookjs/storybook/pull/26010#issuecomment-2136405529)
  - [#issuecomment-2136852474](https://github.com/storybookjs/storybook/pull/26010#issuecomment-2136852474)
  - [#issuecomment-2137047417](https://github.com/storybookjs/storybook/pull/26010#issuecomment-2137047417)
  
> [!NOTE]
> Be aware of Rollup's [virtual module convention](https://vitejs.dev/guide/api-plugin#virtual-modules-convention) for virtual modules adopted by vite: 

- [ ] Updating the tests accordingly
- [ ] Create a simple sandbox example that displays the use of experimental indexers.
  The documentation has [some examples](https://storybook.js.org/docs/api/main-config/main-config-indexers#examples) at the end that you may use so you don't have to create one from scratch.
  Just make sure the path you are using to load the stories is a virtual one or, at least, not the original path

ndelangen committed:
Merge branch 'next' into fix/indexers-ignoring-importPath

tobiasdiez commented:
Note that another alternative to virtual paths is to add certain query segments / "sub parts" to the import url, e.g. `Button.stories.ts?story=abc`. According to https://vitejs.dev/guide/api-plugin#virtual-modules-convention, this can be used if the code in the virtual module is derived from a real file:

> Note that modules directly derived from a real file, as in the case of a script module in a Single File Component (like a .vue or .svelte SFC) don't need to follow this convention. SFCs generally generate a set of submodules when processed but the code in these can be mapped back to the filesystem. 

(the vue vite plugin uses this, eg https://github.com/vitejs/vite-plugin-vue/blob/ac1255989d53bcd98f72f2e6c16946a00402eac6/packages/plugin-vue/src/main.ts#L314-L320)

tobiasdiez commented:
Thinking about this a bit more, would it make sense to write the story index (`storybook-stories.js`) in the form
```ts
const imports = {
   story-id: () => import("import-path")
   ....for each story
}
function importFn(storyId) {
   return await imports[storyId]
}
```
where both `story-id` and `import-path` are obtained via the story indexer. And `import-path` can be any path vite can handle as import (e.g. an absolute path, something with alias or a virtual module).
Then https://github.com/storybookjs/storybook/blob/6388f93b1f3517b8b0dad3dbd05aa5d5313ee5ae/code/core/src/preview-api/modules/store/StoryStore.ts#L141-L147 could be rewritten as
```ts
async loadCSFFileByStoryId(storyId: StoryId): Promise<CSFFile<TRenderer>> {
    const moduleExports = await this.importFn(storyId);

    // We pass the title in here as it may have been generated by autoTitle on the server.
    return this.processCSFFileWithCache(moduleExports, importPath, title);
  }
```

svallory commented:
> Note that another alternative to virtual paths is to add certain query segments / "sub parts" to the import url, e.g. Button.stories.ts?story=abc. According to https://vitejs.dev/guide/api-plugin#virtual-modules-convention, this can be used if the code in the virtual module is derived from a real file:

It's not an alternative to the issue at hand if it requires an existing file.

Also, depending on the language/framework you are creating the plugin for, even if there's a file, you may not be able to process the request before the language/framework plugin does

svallory commented:
> Thinking about this a bit more, would it make sense to write the story index (storybook-stories.js) in the form

@tobiasdiez I think the entire process would benefit from a little cleanup (well, more like a big  refactoring TBH)

The flow is confusing and hard to follow. I suspect over the years the need for optimization, or handling new use cases, led to adding more functions, breaking the process of indexing and loading into so many specialized functions.

But now that you guys have a clearer picture, I think a do-over with a cleaner separation of concerns (introducing a CSFFile class/module for example) would go a long way

tobiasdiez commented:
> > Note that another alternative to virtual paths is to add certain query segments / "sub parts" to the import url, e.g. Button.stories.ts?story=abc. According to https://vitejs.dev/guide/api-plugin#virtual-modules-convention, this can be used if the code in the virtual module is derived from a real file:
> 
> It's not an alternative to the issue at hand if it requires an existing file.
> 
> Also, depending on the language/framework you are creating the plugin for, even if there's a file, you may not be able to process the request before the language/framework plugin does

Sure! That remark was meant to provide an alternative solution as long as this PR is not finished. I would love to see support for `virtual` modules. 
For many purposes (eg your marko plugin ?) the stories should be defined in an existing file, so add a sub part query may work.


> > Thinking about this a bit more, would it make sense to write the story index (storybook-stories.js) in the form
> 
> @tobiasdiez I think the entire process would benefit from a little cleanup (well, more like a big refactoring TBH)
> 
> The flow is confusing and hard to follow. I suspect over the years the need for optimization, or handling new use cases, led to adding more functions, breaking the process of indexing and loading into so many specialized functions.
> 
> But now that you guys have a clearer picture, I think a do-over with a cleaner separation of concerns (introducing a CSFFile class/module for example) would go a long way

From what I have seen in the codebase, I agree with your judgement.

tobiasdiez commented:
Thanks @tmeasday for the nice overview.

One problem with this design is that whatever the indexer returns as the `importPath` is never reaching the builder and thus cannot be used to construct a correct `importFn`.

As an idea, would it work to use the story index `index.json` (via importing it) to build the correct `importFn`? For this maybe the generation of `index.json` needs to be moved to the builder.

tmeasday commented:
> One problem with this design is that whatever the indexer returns as the importPath is never reaching the builder and thus cannot be used to construct a correct importFn.

> As an idea, would it work to use the story index index.json (via importing it) to build the correct importFn? For this maybe the generation of index.json needs to be moved to the builder.

But this is my point. Webpack in particular does not and would not use the generated index to construct it's `importFn`, so although what you are suggesting might work in Vite, it would not have any effect in Webpack.

tobiasdiez commented:
I'm completely ignorant when it comes to webpack, but is there a fundamental issue with having the following dependency chain?
```
stories (via glob) -> indexer (index.json) -> importFn script
```
Then if the stories are changed, the indexer output changes as well; which invalidates `importFn` and triggers the HMR of it I'd assumed that this would work once the indexer output is part of the builder (so that webpack/vite can monitor it).

Another idea would be to call the indexer for the generation of the `importFn` script.

Or how would you envision how the `importFn` makes use of the `importPath` returned by the indexer?

JReinhold commented:
> In general, for files that need compilation before becoming a JS CSF File, there are two options:
> 
> 1. Intercept the transpilation: cut off the language plug-in and instrument the compiler (this is what the Svelte stories plug-in does)
> 
> 2. Use a virtual path: add a virtual path to storybook index, when that path triggers the plug-in, convert it to the real file path, import it and the language plug-in will do its thing and you'll get a JS module (this is what this PR will allow)

> I guess I wonder about what's the issue with the first option?

@tmeasday AFAIK only the second option would provide a way to define multiple separate metas from the same file. Eg if you had a json-file from which you wanted to create 3 components with 10 stories each. Because CSF only allows one component per file, you'd instead generate 3 virtual files from that single json file path.

Although TBH I haven't actually tried returning IndexInputs with different titles from the same file path.

The second option does sound slightly interesting to me, based on the Svelte work I've done. Given you could instead create a virtual CSF file that wraps the original Svelte component I imagine the complexity around transforming an already-internally-transformed file would be smaller. A few cases spring to mind:
- today we have an issue with variable collision, if the generated story export variable names collide with user-imported variables. https://github.com/storybookjs/addon-svelte-csf/issues/167
- we depend on referencing the Svelte-internal component variable. If they changed that variable name in a patch (because it's an internal construct) it would break the addon.
- we have to be careful (and the user too) about which order the Svelte CSF Vite plugin is added, because its transformer _must_ run _after_ the Svelte transformation, but _before_ the experimental Vitest integration addon.

I'd expect all of the above would be solved by moving away from transforming the file inline to generating a wrapper CSF file instead.

svallory commented:
@tmeasday thanks for the overview! 

> I guess I wonder about what's the issue with the first option?

To add to what @JReinhold said about this, the main issue with "intercepting the transpilation" is that the development of a "simple" plugin for storybook suddenly requires knowledge of how the language compiler works.

In the case of Svelte, the plugin needed to run after the Svelte transpilation. In my case, since Marko integrates tightly with the server for extensive runtime optimization (which I believe may also be the case for Qwik and Imba), that wasn't an option. I would need to take control of the Marko compiler, proxying it's integration entirely. Even with deep knowledge of its internals, the implementation would end up fragile and break easily with future changes.

JReinhold commented:
I think we can close this in favor of https://github.com/storybookjs/storybook/pull/30612

While that is also still not done, it's closer to the end result we want.
